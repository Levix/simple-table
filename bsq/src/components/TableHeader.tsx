import { defineComponent, ExtractPropTypes } from "vue";
import { tableProps } from "../types/table"
import { noop } from "lodash";

const headerProps = {
  ...tableProps,
  handleSort: {
    type: Function,
    default: noop
  },
  keyClickMap: {
    type: Object,
    default: {}
  }
} as const

export type HeaderProps = ExtractPropTypes<typeof headerProps>;

export default defineComponent({
  name: 'TableHeader',
  props: headerProps,
  setup(props: HeaderProps) {

    return () => (
      <thead>
        <th>
          {
            props.option.columns.map((item: { title: string; sortable?: boolean; dataKey: string }) => (
              <td key={item.dataKey} class='table-th-td'>
                {item.title}
                {
                  item.sortable &&
                  <span
                    class={`table-th-td__sort-btn status--${props.keyClickMap?.[item?.dataKey]}`}
                    onClick={
                      () => props.handleSort(
                        props.option.data,
                        item.dataKey,
                        props.option.isMultiSort,
                        props.option.isLocalSort
                      )
                    }
                  >{'>'}</span>
                }
              </td>
            ))
          }
        </th>
      </thead>
    )
  }
})