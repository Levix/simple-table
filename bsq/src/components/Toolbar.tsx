import { defineComponent, renderSlot } from "vue";

export default defineComponent({
  name: 'Toolbar',
  setup({}, { emit, slots }) {

    const handleOnKeyDown = (e: KeyboardEvent) => {

      // 处理回车搜索
      if (e.keyCode === 13) {
        emit('search', (e.target as HTMLInputElement).value);
      }
    }

    return () => {

      // 操作栏：左侧插槽
      const slotLeftopr = renderSlot(slots, 'default')

      return (
        <div class="table-wrap-toolbar">
          <div>
            {/* 左侧操作栏插槽 */}
            <slotLeftopr />
          </div>
          <div>
            <input
              type="text"
              class="table-wrap-toolbar__search-input"
              placeholder="可以模糊搜索id"
              onKeydown={e => handleOnKeyDown(e)} />
            <button
              class="table-wrap-toolbar__refresh-btn"
              onClick={() => emit('refresh')}
            >刷新</button>
          </div>
        </div>
      )
    }
  }
})