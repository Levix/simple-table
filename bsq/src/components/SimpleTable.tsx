import { computed, defineComponent, reactive, renderSlot, watch } from "vue"
import { type TableProps, tableProps } from "../types/table"
import type { SortParamsType } from "../types/app"
import { cloneDeep } from 'lodash'
import Toolbar from './Toolbar'
import useSortable from "../hooks/useSortable"
import TableHeader from "./TableHeader"
import TableBody from "./TableBody"
import { Logger } from '../util/logger'
import '/src/styles/index.css'

export default defineComponent({
  name: 'SimpleTable',
  components: {
    Toolbar,
    TableHeader,
    TableBody
  },
  props: tableProps,
  setup(props: TableProps, { emit, slots }) {

    // 同步
    let list = reactive({
      data: cloneDeep(props.option.data)
    });

    let { SORT_PARAMS_MAP, keyClickMap, sortableData, handleSort } = useSortable()

    watch(sortableData, () => {
      list.data = sortableData.value
    })

    const curOptionData = computed(() => props.option.data)

    // 异步
    watch(curOptionData, () => {
      list.data = cloneDeep(curOptionData.value)
    })

    watch(keyClickMap, () => {
      let sortParams = Object.keys(keyClickMap.value).reduce((prev, key) => {
        prev[key] = SORT_PARAMS_MAP[keyClickMap.value[key]]
        return prev;
      }, {} as SortParamsType)
      Logger.info('当前搜索过滤参数：', sortParams)
      emit('handle-sort', sortParams)
    }, {
      deep: true
    })

    return () => {

      // 操作栏：左侧插槽
      const slotLeftopr = renderSlot(slots, 'default')

      return (
        <div class='table-wrap'>
          <toolbar
            onSearch={(val: string) => emit('on-search', val)}
            onRefresh={() => emit('on-refresh')}
          >
            <slotLeftopr />
          </toolbar>

          <table>
            <table-Header
              option={props.option}
              keyClickMap={keyClickMap.value}
              handleSort={handleSort} />

            <table-body
              option={props.option}
              list={list.data} />
          </table>
        </div>
      )
    }
  },
});
