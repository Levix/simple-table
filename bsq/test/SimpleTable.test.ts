import { shallowMount, mount } from '@vue/test-utils';
import { expect, describe, it } from 'vitest';
import SimpleTable from '../src/components/SimpleTable';
import Toolbar from '../src/components/Toolbar';
import TableHeader from '../src/components/TableHeader';
import { tableProps } from './mock';

describe('mount SimpleTable', () => {
  it('renders SimpleTable', () => {
    const wrapper = shallowMount(SimpleTable, {
      props: {
        option: tableProps
      },
    })

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.exists()).toBe(true)
  })

  it('模拟异步渲染：renders SimpleTable', async () => {
    const wrapper = shallowMount(SimpleTable)

    await new Promise((res, rej) => {
      setTimeout(() => {
        res(tableProps)
      }, 1000)
    }).then(res => {
      wrapper.setProps({
        option: res
      })
    });

    expect(wrapper.exists()).toBe(true)
  })

  it('SimpleTable renders default slot', () => {
    const defaultSlot = 'default slot message'
    const wrapper = mount(SimpleTable, {
      slots: {
        default: defaultSlot
      }
    })

    expect(wrapper.html()).toContain(defaultSlot)
  })

  it('Toolbar button refresh click', async () => {
    const wrapper = mount(SimpleTable).findComponent(Toolbar)
    await wrapper.find('.table-wrap-toolbar__refresh-btn').trigger('click')
  })

  it('Toolbar renders default slot', () => {
    const defaultSlot = 'default slot message'
    const wrapper = mount(Toolbar, {
      slots: {
        default: defaultSlot
      }
    })

    expect(wrapper.html()).toContain(defaultSlot)
  })

  it('Toolbar input search', async () => {
    const wrapper = mount(SimpleTable)
      .findComponent(Toolbar)
      .find('.table-wrap-toolbar__search-input')

    await wrapper.setValue('')
    await wrapper.trigger('onKeydown', {
      keyCode: 13
    })
  })

  it('TableHeader title label sort：升序，降序，无序', async () => {
    const wrapper = mount(SimpleTable).findComponent(TableHeader)
    await wrapper.find('.table-th-td__sort-btn').trigger('click')
    await wrapper.find('.table-th-td__sort-btn').trigger('click')
    await wrapper.find('.table-th-td__sort-btn').trigger('click')
  })
})

describe('mount empty SimpleTable', () => {
  const wrapper = mount(SimpleTable)

  it('renders SimpleTable', () => {
    expect(wrapper.exists()).toBe(true)
  })
  
  it('空状态是否存在', () => {
    expect(wrapper.find('.table-th-td__empty').exists()).toBe(true)
  })
})
