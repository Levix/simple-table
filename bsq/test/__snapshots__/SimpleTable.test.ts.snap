// Vitest Snapshot v1

exports[`mount SimpleTable > renders SimpleTable 1`] = `
"<div class=\\"table-wrap\\">
  <toolbar-stub></toolbar-stub>
  <table>
    <table-header-stub option=\\"[object Object]\\" handlesort=\\"(originData, key, isMultiSort = false, isLocalSort = false) => {
    let list = __vite_ssr_import_0__.cloneDeep(originData);
    if (!isMultiSort) {
      keyClickMap.value = __vite_ssr_import_0__.cloneDeep({
        [key]: keyClickMap.value[key] || SORT_MAP.normal
      });
    }
    if (key in keyClickMap.value) {
      keyClickMap.value[key] = (keyClickMap.value[key] + 1) % Object.keys(SORT_MAP).length;
    } else {
      keyClickMap.value[key] = SORT_MAP.asc;
    }
    if (isLocalSort) {
      switch (keyClickMap.value[key]) {
        case SORT_MAP.asc:
          list.sort((a, b) => a[key] - b[key]);
          break;
        case SORT_MAP.desc:
          list.sort((a, b) => b[key] - a[key]);
          break;
        default:
          break;
      }
    }
    sortableData.value = list;
  }\\" keyclickmap=\\"[object Object]\\"></table-header-stub>
    <table-body-stub option=\\"[object Object]\\" list=\\"[object Object],[object Object]\\"></table-body-stub>
  </table>
  <pagination-stub option=\\"[object Object]\\"></pagination-stub>
</div>"
`;
